In [190]:
#http://www.tensorfly.cn/tfdoc/tutorials/mnist_beginners.html
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data


"""
URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] 
certificate verify failed (_ssl.c:749)>

解决办法一：
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

解决办法二：
/Applications/Python 3.6/ReadMe.rtf中提到：

Certificate verification and OpenSSL

**NEW** This variant of Python 3.6 now includes its own private copy of OpenSSL 1.0.2.  
Unlike previous releases, the deprecated Apple-supplied OpenSSL libraries are no longer used.  
This also means that the trust certificates in system and user keychains managed by the 
Keychain Access application and the security command line utility are no longer used as 
defaults by the Python ssl module.  For 3.6.0, a sample command script is included in 
/Applications/Python 3.6 to install a curated bundle of default root certificates from the 
third-party certifi package (https://pypi.python.org/pypi/certifi).  If you choose to 
use certifi, you should consider subscribing to the project's email update service to be
notified when the certificate bundle is updated.

The bundled pip included with the Python 3.6 installer has its own default certificate store 
for verifying download connections.

具体操作：
执行/Applications/Python 3.6/Install Certificates.command
"""

mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)
mnist

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


Datasets(train=<tensorflow.contrib.learn.python.learn.datasets.mnist.DataSet object at 0x24c9214e0>, validation=<tensorflow.contrib.learn.python.learn.datasets.mnist.DataSet object at 0x276c2cb00>, test=<tensorflow.contrib.learn.python.learn.datasets.mnist.DataSet object at 0x276c4f320>)

In [191]:
#设置变量
x=tf.placeholder(tf.float32,[None,784])
y=tf.placeholder(tf.float32,[None,10])

Weights=tf.Variable(tf.zeros([784,10]))
biases=tf.Variable(tf.zeros([10]))

x,y,Weights,biases

(<tf.Tensor 'Placeholder_53:0' shape=(?, 784) dtype=float32>,
 <tf.Tensor 'Placeholder_54:0' shape=(?, 10) dtype=float32>,
 <tf.Variable 'Variable_54:0' shape=(784, 10) dtype=float32_ref>,
 <tf.Variable 'Variable_55:0' shape=(10,) dtype=float32_ref>)

In [192]:
#构造模型
prediction=tf.nn.softmax(tf.matmul(x,Weights)+biases)

#loss
# loss=tf.square(prediction-y)
# loss=tf.reduce_mean(tf.square(prediction-y))
# loss=tf.nn.softmax_cross_entropy_with_logits_v2(labels=y,logits=prediction)
loss=tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits_v2(labels=y,logits=prediction))

train_step=tf.train.GradientDescentOptimizer(0.01).minimize(loss)

In [193]:
#评估
#correct_prediction是一个bool list
correct_prediction=tf.equal(tf.argmax(y,1),tf.argmax(prediction,1))
#准确率
"""
tf.cast:转化类型 True->1.0;False->0.0
tf.reduce_mean:求平均值，即准确率
"""
accuracy=tf.reduce_mean(tf.cast(correct_prediction,tf.float32))

In [194]:
sess=tf.Session()
sess.run(tf.global_variables_initializer())

In [195]:
#训练
for i in range(30000):
    batch_xs, batch_ys = mnist.train.next_batch(100)
    sess.run(train_step, feed_dict={x: batch_xs, y: batch_ys})
    if i%2000==0:
        res=sess.run(accuracy,feed_dict={x:mnist.test.images,y:mnist.test.labels})
        print(res)

0.3013
0.7443
0.8034
0.8135
0.8196
0.8239
0.826
0.8298
0.86
0.8751
0.8853
0.8917
0.8956
0.897
0.8989


# 不同loss对应的训练效果

#### loss=tf.square(prediction-y)

0.1428
0.9214
0.9255
0.9269
0.927
0.9287
0.9301
0.928
0.9309
0.9285
0.9268
0.9283
0.931
0.9306
0.93
***
#### loss=tf.reduce_mean(tf.square(prediction-y))

0.5336
0.6663
0.7053
0.7565
0.7869
0.8228
0.8403
0.8484
0.8586
0.864
0.8681
0.8707
0.8736
0.8759
0.8781
***
#### loss=tf.nn.softmax_cross_entropy_with_logits(labels=y,logits=prediction)

0.1872
0.9199
0.9249
0.9265
0.9264
0.927
0.9271
0.9275
0.9276
0.9288
0.929
0.9286
0.928
0.9292
0.9296
***
#### loss=tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y,logits=prediction))

0.267
0.7398
0.8028
0.8136
0.8203
0.8238
0.8264
0.8301
0.8608
0.876
0.8856
0.892
0.8955
0.8973
0.899